lambdas/end-mitigation-journey/src/main/java/uk/gov/di/ipv/core/endmitigationjourney/EndMitigationJourneyHandler.java

```diff
@@ -5,21 +5,215 @@
+            return ApiGatewayResponseGenerator.proxyJsonResponse(
+                    HttpStatus.SC_INTERNAL_SERVER_ERROR, ErrorResponse.FAILED_TO_GET_STORED_CIS);
+        }
+
+        return ApiGatewayResponseGenerator.proxyJsonResponse(HttpStatus.SC_OK, JOURNEY_NEXT);
+    }
+
+    private Optional<List<String>> validateMitigationJourney(
+            List<String> credentials,
+            ContraIndicatorItem ciItem,
+            String mitigationId,
+            ConfigurationService configurationService)
+            throws UnknownMitigationJourneyException {
+        // check mitigation id and call vorrect validation method
+        if (mitigationId.equals(MJ01)) {
+            return Mj01Validation.validateJourney(credentials, ciItem, configurationService);
+        }
+        throw new UnknownMitigationJourneyException(
+                String.format("Unknown mitigation journey id: %s", mitigationId));
+    }
+
+    private void endMitigationJourney(
+            IpvSessionItem ipvSessionItem,
+            IpvSessionService ipvSessionService,
+            String mitigationJourneyId) {
+        List<ContraIndicatorMitigationDetailsDto> contraIndicatorMitigationDetails =
+                new ArrayList<>(ipvSessionItem.getContraIndicatorMitigationDetails());
+
+        contraIndicatorMitigationDetails.forEach(
+                contraIndicatorMitigationDetailsDto ->
+                        contraIndicatorMitigationDetailsDto
+                                .getMitigationJourneys()
+                                .forEach(
+                                        mitigationJourneyDetails -> {
+                                            if (mitigationJourneyDetails
+                                                    .getMitigationJourneyId()
+                                                    .equals(mitigationJourneyId)) {
+                                                mitigationJourneyDetails.setComplete(true);
+                                            }
+                                        }));
+
+        ipvSessionItem.setContraIndicatorMitigationDetails(contraIndicatorMitigationDetails);
+        ipvSessionService.updateIpvSession(ipvSessionItem);
+    }
+
+    private void submitMitigatingVcs(
+            List<String> mitigatingVcs, IpvSessionItem ipvSessionItem, String ipAddress) {
+        try {
+            ciStorageService.submitMitigatingVcList(
+                    mitigatingVcs,
+                    ipvSessionItem.getClientSessionDetails().getGovukSigninJourneyId(),
+                    ipAddress);
+        } catch (CiPostMitigationsException e) {
+            StringMapMessage mapMessage =
+                    new StringMapMessage()
+                            .with(
+                                    "message",
+                                    "Failed to send mitigation request to the CI storage system")
+                            .with("reason", e.getMessage());
+            LOGGER.error(mapMessage);
```

Comment:
Shouldn't we be failing the journey here because CI storage has probably failed to mark the CIs as mitigated? (or is the intention to just swallow the error for now?)

Reply:
Yeah for now I thought it would be better to just log the error and move on. This way it means that the lambda will still mark this particular mitigation journey as complete so that we don't get stuck in a loop retrying the same mitigation journey. And if the CI Storage failed for whatever reason then we can still attempt an identity verification with CI scoring etc